### PR TITLE
Reduce code duplication

### LINKED ISSUE #129: Refactoring of GitHub/GitLab Behavior and Service classes
While looking at the source code to implement the fix for #127, I noticed that there is a fairly large amount of duplicated code between the GitHub and GitLab providers (Service / Behavior), most of which could be relatively easily factored out into a common parent abstract class.

If you guys are willing to consider such contributions, I will gladly submit a pull request with proposed changes.

Please let me know if you're interested.

### PR DESCRIPTION
**Pull request by dregad**
Refactoring of the code to avoid duplication between GitHub and GitLab classes, with the introduction of 3 new abstract "Git Provider" classes for the GitHub and GitLab Service, Status and Behavior.

Fixes #129

### CODE REVIEW

**Comment by Klap-in** [comment]
If you change a service class, the service configuration must be changed accordingly.

I have setup my local instance again. I'm now reviewing and testing, but I have not much time per day. So will take some other days.

**Comment by dregad** [comment]
> If you change a service class, the service configuration must be changed accordingly.

Sorry about that, I know nothing about Symfony development. Thanks for fixing it !

> I have setup my local instance again. I'm now reviewing and testing, but I have not much time per day. So will take some other days.

Great. Take your time there is no urgency - I was just pinging you guys to make sure it was not forgotten.

Let me know if there is anything I can or should do.

File: src/Services/GitHostingProviderService.php [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    /**
+     * Hosting Provider identifier.
+     * Must be defined by child class.
+     */
+    protected string $provider;
+
+    /**
+     * Hosting provider-specific Exception to be thrown.
+     */
+    protected string $exceptionType;
```
Comment (Klap-in) [inline src/Services/GitHostingProviderService.php]: Is this really needed? It is only used for the exception on getting the user and repo name. What about renaming the exception to 'GitHostingProviderServiceException' or ´GitHostingServiceException' or .. ? 

Note that GitHubServiceException and GitLabServiceException are mentioned quite a bit through the code. Feels as duplication as well.  Also the error reporter distincts them. But sent almost the same email. In the email the entire git url is mentioned, so in general it is clear if github/gitlab is meant. But GitLab and GitHub are mentioned in text as well, this must be generalized or the name must be set via the exception or something like that. We have to check that the logging is unique enough to prevent potential confusion if repositories are added to the translation tool with same usernames and reponames for github and gitlab.

File: src/Services/GitHostingProviderService.php [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        return $pool;
+    }
+
+    /**
+     * Create fork in our Hosting Provider account
+     *
+     * @param string $url URL to create the fork from
+     * @return string Git URL of the fork
+     */
+    abstract public function createFork(string $url): string;
```
Comment (Klap-in) [inline src/Services/GitHostingProviderService.php:78]: not all thrown exceptions are mentioned. So far I tried to mention all throws, to make it easier to remember how and where they are caught.

File: src/Services/Repository/Behavior/GitHubBehavior.php [hunk 3]
```diff
@@ -2,150 +2,7 @@
...
-     * @return array{count: int, listURL: string, title: string}
-     *
-     * @throws GitHubServiceException
-     */
-    public function getOpenPRListInfo(RepositoryEntity $repository, LanguageNameEntity $language): array
-    {
-        return $this->api->getOpenPRListInfo($repository->getUrl(), $language->getCode());
-    }
-
+    protected string $remote = 'github';
```
Comment (Klap-in) [inline src/Services/Repository/Behavior/GitHubBehavior.php]: What about generalizing further? e.g. using a more general name such as 'remote' or 'githostingprovider', etc.

**Review by Klap-in — COMMENTED** [review]
I made another simplification. Further some questions, could you have a look?
This interaction with github/gitlab is most annoying part to test, still to do.

File: src/Services/GitHostingProviderService.php (hunk 1 above)
Comment (dregad) [reply src/Services/GitHostingProviderService.php]: > Is this really needed? 

Maybe not. Bear in mind that I am not very familiar with the code base. 

> What about renaming the exception to 'GitHostingProviderServiceException' or ´GitHostingServiceException'

Sounds like a good idea - I thought about it also but I was not sure about the implications and I did not want to make too far-reaching changes.

**Comment by dregad** [comment]
Looks OK to me, but I can't do much more than a look at the code - can't test.

**Comment by dregad** [comment]
Just wondering what the status of this was, can it be merged ? Is there anything needed on my end ?

**Comment by Klap-in** [comment]
I started at that time with reactivating the development environment and the gitlab/GitHub instances for development. Some things expired there. In the past I had several issues when testing in the actual circumstances, so I prefer to test first before employing, because edge cases will almost not occur in real and are then difficult to recognize. Sorry for my very slow pace! But unfortunately I have limited time available, what makes this a multiple day effort for me.

**Comment by Klap-in** [comment]
I merge this. I tested in multiple phases in a couple of weeks, so I hope I did not lost the overview. I tried to fix the issues you opened as well.
